llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -0,0 +1,320 @@
+  // If we don't have a dedicated preheader that is fallthrough, don't even
+  // bother.
+  Preheader = AIELoopUtils::getDedicatedFallThroughPreheader(LoopBlock);
+  if (!Preheader) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: No fallthrough preheader\n");
+    return false;
+  }
+
+  // 3. We need to know how to update the tripcount. We check whether
+  // the tripcount is pristine, otherwise the loop may have been software
+  // pipelined before and we can't trust min itercount metadata.
+  // Return on investment is probably low anyway.
+  const bool Pristine = true;
+  for (auto &MI : reverse(*Preheader)) {
+    if (TII->isZOLTripCountDef(MI, Pristine)) {
+      TripCountDef = &MI;
+      break;
+    }
+  }
+  if (!TripCountDef) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: No tripcount def\n");
+    return false;
+  }
+
+  // 4. We need to peel stages and be left with a positive tripcount.
+  // This is just a minimum check to save useless work; the real stage
+  // count is checked before accepting the schedule.
+  auto ParsedMinTripCount = AIELoopUtils::getMinTripCount(LoopBlock);
+  if (!ParsedMinTripCount) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: No min tripcount\n");
+    return false;
+  }
+  MinTripCount = *ParsedMinTripCount;
+  if (MinTripCount < 2) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: min tripcount < 2\n");
+    return false;
+  }
+
+  return true;
+}
+
+// This assigns Cycle of SU, Earliest of its predecessors and Earliest of
+// the next instance of SU.
+void PostPipeliner::scheduleNode(SUnit &SU, int Cycle) {
+  LLVM_DEBUG(dbgs() << "PostPipeline " << SU.NodeNum << " in cycle " << Cycle
+                    << "\n");
+  Info[SU.NodeNum].Cycle = Cycle;
+  for (auto &Dep : SU.Succs) {
+    int Latency = Dep.getSignedLatency();
+    auto *Succ = Dep.getSUnit();
+    if (Succ->isBoundaryNode()) {
+      continue;
+    }
+    int SNum = Succ->NodeNum;
+    Info[SNum].Earliest = std::max(Info[SNum].Earliest, Cycle + Latency);
+    LLVM_DEBUG(dbgs() << "   Earliest(" << SNum << ") to "
+                      << Info[SNum].Earliest << "\n");
+  }
+
+  int Next = SU.NodeNum + NInstr;
+  if (Next < NTotalInstrs) {
+    Info[Next].Earliest = std::max(Info[Next].Earliest, Cycle + II);
+  }
+}
+
+// Check resources. We only insert at the position modulo II. Since we insert
+// all iterations separately, the resources that wrap around accumulate in the
+// overflow area, causing conflicts when inserting future iterations
+int PostPipeliner::fit(MachineInstr *MI, int Earliest, int NTries, int II) {
+  for (int C = Earliest; C < Earliest + NTries; C++) {
+    int Mod = C % II;
+    LLVM_DEBUG(dbgs() << "   at " << C << " (" << Mod << ")\n");
+    if (!HR.checkConflict(Scoreboard, *MI, -Depth + Mod)) {
+      LLVM_DEBUG(dbgs() << "    Success\n");
+      return C;
+    }
+  }
+  LLVM_DEBUG(dbgs() << "    Fail\n");
+
+  return -1;
+}
+
+bool PostPipeliner::scheduleFirstIteration() {
```

Comment:
It is actually the first iteration of the non-pipelined loop, and it will probably span all stages of the schedule, though not have instructions in all stages. If we take a simple load+store loop, the load will be in stage 0, the store in stage 7.  With an II of 1, the second iteration will be in stage 1 and stage 8, the third in stage 2 and stage 9, etc
Each next iteration will be shifted by exactly II by the constraints we apply in scheduleOtherIterations.